ipv8/messaging/deprecated/encoding.py

```diff
@@ -93,9 +104,9 @@ def _a_encode_dictionary(values, mapping):
     {'foo':'bar', 'moo':'milk'} --> ['2', 'd', '3', 's', 'foo', '3', 's', 'bar', '3', 's', 'moo', '4', 's', 'milk']
     """
     assert isinstance(values, dict), "VALUE has invalid type: %s" % type(values)
-    encoded = [str(len(values)).encode("UTF-8"), "d"]
+    encoded = [str(len(values)).encode("UTF-8"), b"d"]
     extend = encoded.extend
-    for key, value in sorted(values.items()):
```

Comment:
Oh crap we need this, but: https://portingguide.readthedocs.io/en/latest/comparisons.html#unorderable-types

Reply:
@devos50 this is a big problem, TrustChain will never work on Python 3 like this:

The problem:
 - When we load a block from the database we repack it, which will likely alter the ordering of the transaction.
 - We depend on the `sorted()` producing the same sorting of differently typed inputs.
 - Given the previous two points, the signature does not match the newly packed block.

If we do not fix this now, we are guaranteed a hard fork of TrustChain later.

Reply:
Right... this is a problem indeed.

In #282, I'm changing quite some stuff regarding TrustChain so something that would fix it (and requires a protocol break) can go in that PR. Note that I'm not doing a full block wipe (yet) but I have this planned for the actual release of Tribler 7.1.

Another problem, though unrelated to this PR, is that encoding/decoding of transactions takes quite some time and this is a good opportunity to explore a faster approach.

Reply:
I would prefer to get rid of `encoding.py` since it's slow and yet another serialization protocol we have to support. As far as I know, TrustChain is the only component using functionality in `encoding.py`.

Using the payload system is a good idea: it's well tested and fast. I think however that we have to extend it to support nested dictionaries (like I'm using in the market).

Reply:
Anyway, it should be pretty easy to make a `Payload` directive for the serializer. I'll try to get that up and running today.

Reply:
As expected adding `sorted()` here again makes the Tribler PR pass: https://github.com/Tribler/tribler/pull/3899